pybamm/meshes/one_dimensional_submeshes.py

```diff
@@ -30,10 +31,62 @@ def near(x, point, tol=3e-16):
                 return abs(x - point) < tol
 
             for tab in ["negative", "positive"]:
-                if near(tabs[tab]["z_centre"], 0):
+                tab_location = tabs[tab]["z_centre"]
+                if near(tab_location, 0):
                     self.tabs[tab + " tab"] = "left"
-                elif near(tabs[tab]["z_centre"], l_z):
+                elif near(tab_location, l_z):
                     self.tabs[tab + " tab"] = "right"
+                else:
+                    raise pybamm.GeometryError(
+                        """{} tab located at {}, but must be at either 0 or {}
+                        (in dimensionless coordinates).""".format(
+                            tab, tab_location, l_z
+                        )
+                    )
+
+
+class MeshGenerator1D(MeshGenerator):
+    """
+    A class to generate a submesh on a 1D domain.
+
+    Parameters
+    ----------
+
+    submesh_type: str, optional
+        The type of submeshes to use. Can be "Uniform", "Exponential", "Chebyshev"
+        or "User". Default is "Uniform".
+    submesh_params: dict, optional
+        Contains any parameters required by the submesh.
+
+    **Extends**: :class:`pybamm.MeshGenerator`
+    """
+
+    def __init__(self, submesh_type="Uniform", submesh_params=None):
```

Comment:
submesh_type could be the class itself e.g. Uniform1DSubMesh. And all Submesh types can have inputs lims, npts, tabs, submesh_params. Then we just have: 
```python
def __call__(self, lims, npts, tabs=None)
return self.submesh_type(lims, npts, tabs, self.submesh_params)
``` 
Just to have fewer add/maintain if new meshes are added in the future